Classes/Bookmark/BookmarksViewController.swift

```diff
@@ -0,0 +1,218 @@
+//
+//  BookmarksViewController.swift
+//  Freetime
+//
+//  Created by Rizwan on 18/10/17.
+//  Copyright © 2017 Ryan Nystrom. All rights reserved.
+//
+
+import UIKit
+
+class BookmarksViewController:
+    UITableViewController,
+    UISearchBarDelegate,
+    PrimaryViewController,
+TabNavRootViewControllerType {
+  
+    private let client: GithubClient
+    private let cellIdentifier = "bookmark_cell"
+    private let bookmarkStore = BookmarksStore.shared
+    private let searchBar = UISearchBar()
+    private var filterdBookmarks: [BookmarkModel]? = nil
+    
+    // MARK: Init
+    
+    init(client: GithubClient) {
+        self.client = client
+        super.init(nibName: nil, bundle: nil)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    // MARK: View Life Cycle
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        searchBar.delegate = self
+        searchBar.placeholder = NSLocalizedString("Search", comment: "")
+        searchBar.tintColor = Styles.Colors.Blue.medium.color
+        searchBar.backgroundColor = .clear
+        searchBar.searchBarStyle = .minimal
+        searchBar.sizeToFit()
+        tableView.tableHeaderView = searchBar
+        
+        navigationItem.rightBarButtonItem = UIBarButtonItem(
+            title: NSLocalizedString("Clear All", comment: ""),
+            style: .plain,
+            target: self,
+            action: #selector(BookmarksViewController.onClearAll(sender:))
+        )
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        rz_smoothlyDeselectRows(tableView: tableView)
+        tableView.reloadData()
+        updateRightBarItem()
+    }
+    
+    override func viewWillLayoutSubviews() {
+        super.viewWillLayoutSubviews()
+    }
+    
+    // MARK: Private API
+    
+    func updateRightBarItem() {
+        navigationItem.rightBarButtonItem?.isEnabled = bookmarkStore.bookmarks.count > 0
+    }
+    
+    @objc private func onClearAll(sender: UIBarButtonItem) {
+        let alert = UIAlertController.configured(
+            title: NSLocalizedString("Bookmarks", comment: ""),
+            message: NSLocalizedString("Clear all bookmarks?", comment: ""),
+            preferredStyle: .alert
+        )
+        
+        alert.addActions([
+            AlertAction.clearAll({ [weak self] _ in
+                self?.bookmarkStore.clear()
+                self?.tableView.reloadData()
+                self?.updateRightBarItem()
+            }),
+            AlertAction.cancel()
+        ])
+        
+        present(alert, animated: true)
+    }
+    
+    func filter(query: String?) {
+        if let query = query {
+            filterdBookmarks = filtered(array: bookmarkStore.bookmarks, query: query)
+        }
+        else {
+            filterdBookmarks = nil
```

Comment:
Does it make sense for the filteredBookmarks to be nullable? If we have a `nil` query, shouldn't `filteredBookmarks` be an empty array? What would happen now if the array is in fact empty instead of `nil`?